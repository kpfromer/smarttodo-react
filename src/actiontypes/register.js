
export const REQUEST_REGISTER = 'register/REQUEST_REGISTER';
export const RECEIVE_REGISTER = 'register/RECEIVE_REGISTER';
export const FAILURE_REGISTER = 'register/FAILURE_REGISTER';