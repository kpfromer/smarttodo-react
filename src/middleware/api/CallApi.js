import { API_ROOT } from "./api";
import axios from 'axios';

export default (endpoint, method, mapResponse = value => value, data = undefined, token = undefined) => {

  const fullUrl = endpoint.includes(API_ROOT) ? endpoint : API_ROOT + endpoint;

  return axios({
    url: fullUrl,
    method,
    headers: {
      'Content-Type': 'application/json',
      'Authorization': token ? `Bearer ${token}` : undefined
    },
    data
  })
    .then(response => mapResponse(response.data))
    .catch(error => Promise.reject(error));
};
