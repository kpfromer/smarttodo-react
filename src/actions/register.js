import { CALL_API } from "../middleware/api/api";
import { FAILURE_REGISTER, RECEIVE_REGISTER, REQUEST_REGISTER } from "../actiontypes/register";

export const tryRegister = (username, password, email, firstName, lastName) => ({
  [CALL_API]: {
    types: [
      REQUEST_REGISTER,
      RECEIVE_REGISTER,
      FAILURE_REGISTER
    ],
    method: 'POST',
    endpoint: '/register',
    data: {
      username,
      password,
      email,
      firstName,
      lastName
    },
    authenticate: false
  }
});
