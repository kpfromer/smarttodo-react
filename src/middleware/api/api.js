import CallApi from './CallApi';

export const API_ROOT = 'http://localhost:3001/v1';
export const CALL_API = 'call/CALL_API';

const CREATE_TYPES = [
  'POST',
  'PUT'
];
const VALID_TYPES = [
  ...CREATE_TYPES,
  'GET',
  'DELETE'
];

export default ({ getAuthFromState } = { getAuthFromState: val => val }) => store => next => action => {
  const callAPI = action[CALL_API];

  if (typeof callAPI === 'undefined') {
    return next(action);
  }

  let { endpoint } = callAPI;
  const { types, mapResponse, method, data, authenticate = true } = callAPI;

  if (typeof endpoint === 'function') {
    endpoint = endpoint(store.getState())
  }

  if (typeof endpoint !== 'string') {
    throw new Error('Specify a string endpoint URL.')
  }
  if (!Array.isArray(types) || types.length !== 3) {
    throw new Error('Expected an array of three action types.')
  }
  if (!types.every(type => typeof type === 'string')) {
    throw new Error('Expected action types to be strings.')
  }

  if (!VALID_TYPES.includes(method)) {
    throw new Error(`Invalid call api method: ${method}`);
  }

  if (CREATE_TYPES.includes(method) && !data) {
    throw new Error(`Method: ${method} requires data. Data is currently: ${data}`);
  }

  const actionWith = data => {
    const finalAction = { ...action, ...data };
    delete finalAction[CALL_API];
    return finalAction;
  };

  let token;

  if (authenticate) {
    token = getAuthFromState(store.getState());
  }

  const [requestType, successType, failureType] = types;
  next(actionWith({ type: requestType }));

  return CallApi(endpoint, method, mapResponse, data, token).then(
    response => next(actionWith({
      response,
      type: successType
    })),
    error => next(actionWith({
      type: failureType,
      error: error.message || 'Something bad happened'
    }))
  )
};
