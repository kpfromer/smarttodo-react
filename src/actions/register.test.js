import * as register from './register';
import { CALL_API } from "../middleware/api/api";

describe('register actions', () => {
  describe('tryRegister', () => {
    let username, password, email, firstName, lastName;

    beforeEach(() => {
      username = 'exampleUsername';
      password = 'examplePassword';
      email = 'exampleEmail@email.org';
      firstName = 'firstIsMyName';
      lastName = 'lastIsMyName';
    });

    it('creates an action to POST user details', () => {
      const action = register.tryRegister(username, password, email, firstName, lastName);
      expect(action[CALL_API]).toMatchSnapshot();
    });
  });
});