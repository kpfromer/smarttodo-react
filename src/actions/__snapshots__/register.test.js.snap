// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register actions tryRegister creates an action to POST user details 1`] = `
Object {
  "authenticate": false,
  "data": Object {
    "email": "exampleEmail@email.org",
    "firstName": "firstIsMyName",
    "lastName": "lastIsMyName",
    "password": "examplePassword",
    "username": "exampleUsername",
  },
  "endpoint": "/register",
  "method": "POST",
  "types": Array [
    "register/REQUEST_REGISTER",
    "register/RECEIVE_REGISTER",
    "register/FAILURE_REGISTER",
  ],
}
`;
